xof.go

```diff
@@ -0,0 +1,37 @@
+package kyber
+
+import (
+	"crypto/cipher"
+	"io"
+)
+
+// An XOF is an extendable output function, which is a cryptographic
+// primitive that can take arbitrary input like a hash, and then create
+// a stream of output, up to a limit determined by the size of the internal
+// state of the XOF.
+type XOF interface {
+	// Write absorbs more data into the hash's state. It panics if called
+	// after Read. Use Reseed() to
+	io.Writer
+
+	// Read reads more output from the hash. It returns io.EOF if the limit
+	// has been reached.
+	io.Reader
+
+	cipher.Stream
+
+	// Clone returns a copy of the XOF in its current state.
+	Clone() XOF
+
+	// KeySize is the number of bytes that must be written to the XOF before
+	// it is ready to give secure output.
+	KeySize() int
+
+	// Reseed makes an XOF writeable again after it has been read from.
+	Reseed()
+}
+
+// An XOFFactory is an interface that can be mixed in to local suite definitions.
+type XOFFactory interface {
+	XOF([]byte) XOF
```

Comment:
A small comment explaining what is expected as the argument would be nice !

Reply:
One question related to the HashFactory and/or XOFFactory topic:  Is the "XOF version" of BLAKE2 domain-separated from the "hash version" of BLAKE2?  Meaning that calling BLAKE2-HASH("hello world") will produce a different output from calling BLAKE2-XOF("hello world", HASH_SIZE)?  I suspect it probably is domain-separated in this way (it's standard good crypto practice), but I haven't looked at the spec.  

Certainly in the case of SHA3 (hashes) versus SHAKE (XOFs), I believe these are domain separated so that no SHA3 hash can have a domain collision with any SHAKE or vice versa.

At any rate, for flexibility and binary protocol compatibility we'll still need to be able to distinguish from a "hash function used as a hash" and an "XOF used as a hash", for domain-separation purposes if nothing else.  For this reason, I suspect it's likely to be still useful to have both HashFactory and XOFFactory type interfaces, separately.  In fact a single "BLAKE2 factory" object might satisfy both HashFactory and XOFFactory interfaces, but produce instances of BLAKE2 that use different domain-separation values, one for "BLAKE2 as a hash" and one for "BLAKE2 as an XOF."  

So to implement an existing protocol that calls for a hash (e.g., BLAKE2-as-hash or SHA3), we may need the HashFactory interface to get at that, to produce a BLAKE2 instance on the proper "side" of the domain-separation fork for purposes of that particular protocol.

Reply:
Mhhh right. That's indeed a problem if we want to use XOF as a drop-in replacement for hash functions. Further tests are needed indeed.

Reply:
Another problem I had when trying to implement this was that hash.Hash defines Sum as "Sum appends the current hash to b and returns the resulting slice. It does not change the underlying hash state." The obvious way to implement Sum on an XOF is to make a new []byte with length equal to the correct length for a hash of that type, then fill it with Read. But issuing a Read on an XOF changes the state of the XOF, which is against the hash.Hash spec.

So I'm going to stop working on removing HashFactory.

Reply:
Oh right - I'd forgotten about that particular behavior of Sum(), but now that you mention it I've encountered it before and the (minor) implementation challenges it creates here and there.  

One way to implement Sum() "generically", which I might have done at some point in the old Cipher code-base, is for Sum() to first Clone() the Cipher/XOF object, then call Read() on the clone, and otherwise just throw it away.  Not ideal in terms of either cleanness or memory-allocation efficiency though (memory allocations are moderately expensive in Go).

Reply:
Unless you really push, I'm ready to stop working on this. The Cipher->XOF change is a huge improvement in the public API. It seems likely that trying to unify XOF and Hash will make XOF much dirtier, and in the likely case that some protocol needs vanilla standards-compliant hashing, they can't use XOF for that anyway.

If you agree, then let's focus reviews on the details of just what's here in this PR so that early next week I can check this in and move on to the next big cleanup.

I could especially use help understanding if Reseed is doing something cryptographically literate, and how to document KeySize, and how it is set for both impls.

(BTW: I have tested onet's kyber branch against this "Cipher-less" kyber, and there is no breakage.)